### PR TITLE
Remove refs from Pat slices

### PR DESCRIPTION
**Pull request by camsteffen**
Changes `PatKind::Or(&'hir [&'hir Pat<'hir>])` to `PatKind::Or(&'hir [Pat<'hir>])` and others. This is more consistent with `ExprKind`, saves a little memory, and is a little easier to use.

### CODE REVIEW

**Comment by camsteffen** [comment]
@rust-timer queue

**Comment by camsteffen** [comment]
Tried to remove another ref but it looks like I can't so reverted.

File: compiler/rustc_ast_lowering/src/pat.rs [hunk 1]
```diff
@@ -10,7 +10,11 @@ use rustc_span::symbol::Ident;
 use rustc_span::{source_map::Spanned, Span};
 
 impl<'a, 'hir> LoweringContext<'a, 'hir> {
-    crate fn lower_pat(&mut self, mut pattern: &Pat) -> &'hir hir::Pat<'hir> {
+    crate fn lower_pat(&mut self, pattern: &Pat) -> &'hir hir::Pat<'hir> {
+        self.arena.alloc(self.lower_pat_mut(pattern))
+    }
+
+    crate fn lower_pat_mut(&mut self, mut pattern: &Pat) -> hir::Pat<'hir> {
```
Comment (oli-obk) [inline compiler/rustc_ast_lowering/src/pat.rs:17]: The `_mut` confused me a bit. Did you mean "owned and not arena allocated"? I think we usually go for something like `_inner`, but other suggestions are also welcome.

Comment (camsteffen) [reply compiler/rustc_ast_lowering/src/pat.rs:17]: I agree it's not very intuitive. I think it's consistent with existing code at least. I think of it as "use this if you want to modify the output or put it in a (mutable) container".

Comment (oli-obk) [reply compiler/rustc_ast_lowering/src/pat.rs:17]: oh, there's precedent for these postfixes? still confusing, but yea, no need to change that in this pr

**Comment by GuillaumeGomez** [comment]
There is a conflict to be resolved.

@bors: r-

**Pull request merged**